api/unstable/ktls.h

```diff
@@ -0,0 +1,115 @@
+ *
+ * See above for the limitations on when kTLS can be enabled. Additionally,
+ * s2n_connection_ktls_enable_send must be called after the handshake completes.
+ * It may be called after some application data is sent and received without kTLS,
+ * but there must be no pending application data that requires flushing. If these
+ * requirements are not met, enabling kTLS will fail with an error.
+ *
+ * After kTLS is enabled for sending, s2n_send, s2n_sendv, and s2n_sendv_with_offset
+ * will use kTLS. kTLS should result in memory and CPU savings. s2n_sendfile will
+ * also become available.
+ *
+ * For applications using kTLS to avoid copying or allocating memory, s2n_sendv
+ * should be preferred over s2n_sendv_with_offset. For s2n_sendv_with_offset,
+ * s2n-tls may need to copy the provided iovec array to apply the offset, and may
+ * need to allocate memory to copy large (>16) iovec arrays.
+ *
+ * If kTLS is enabled for sending, s2n_connection_get_wire_bytes_out will always
+ * return 0 instead of an accurate count.
+ *
+ * @warning Due to the uncertainty around kTLS support, the signature of this
+ * method is likely to change before kTLS is marked as stable.
+ *
+ * @param conn A pointer to the connection.
+ * @returns S2N_SUCCESS if kTLS is successfully enabled. If kTlS is not successfully
+ * enabled, returns S2N_FAILURE but the connection may proceed without kTLS.
+ */
+S2N_API int s2n_connection_ktls_enable_send(struct s2n_connection *conn);
+
+/**
+ * Enables receiving using kTLS on a given connection.
+ *
+ * See above for the limitations on when kTLS can be enabled. Additionally,
+ * s2n_connection_ktls_enable_recv must be called after the handshake completes.
+ * It may be called after some application data is sent and received without kTLS,
+ * but there must be no buffered application data that requires draining. If these
+ * requirements are not met, enabling kTLS will fail with an error.
+ *
+ * After kTLS is enabled for receiving, s2n_recv will use kTLS. This may result
+ * in memory and CPU savings, but currently will still buffer and copy application data.
+ * We will further optimize s2n_recv for kTLS in the future.
+ *
+ * If kTLS is enabled for receiving, s2n_connection_get_wire_bytes_in will always
+ * return 0 instead of an accurate count.
+ *
+ * @warning Due to the uncertainty around kTLS support, the signature of this
+ * method is likely to change before kTLS is marked as stable.
+ *
+ * @param conn A pointer to the connection.
+ * @returns S2N_SUCCESS if kTLS is successfully enabled. If kTlS is not successfully
+ * enabled, returns S2N_FAILURE but the connection may proceed without kTLS.
+ */
+S2N_API int s2n_connection_ktls_enable_recv(struct s2n_connection *conn);
+
+/**
+ * Sends the contents of a file as application data.
+ *
+ * s2n_sendfile should be more efficient than s2n_send because the copy between
+ * the file and the write socket happens inside the kernel.
```

Comment:
nit:

```suggestion
 * Since data is encrypted in the kernel, s2n_sendfile copies the data directly to the kernel and
 * eliminate a copy to userspace. Compared to s2n_send, s2n_sendfile should be more efficient.
```

> copy between the file and the write socket happens inside the kernel

Its not possible to use sendfile without kTLS so i'll talk about send instead. With send we copy data from disk -> userspace -> kernel -> socket. So there is always a copy from the kernel to the socket. What sendfile really gives us is eliminating a copy to usespace so I would rephrase to emphasize that.

Reply:
This wording is based on the [sendfile documentation](https://man7.org/linux/man-pages/man2/sendfile.2.html):
> sendfile() copies data between one file descriptor and another. **Because this copying is done within the kernel**, sendfile() is more efficient than the combination of [read(2)](https://man7.org/linux/man-pages/man2/read.2.html) and [write(2)](https://man7.org/linux/man-pages/man2/write.2.html), which would require transferring data to and from user space.

I don't think your suggested change is accurate.

> Since data is encrypted in the kernel,

Data is encrypted in the kernel in both s2n_send and s2n_sendfile when ktls is used. That isn't relevant to a discussion of which is more efficient.

> s2n_sendfile copies the data directly to the kernel and eliminate[s] a copy to userspace.

This phrasing implies that s2n-tls is copying something to the kernel. Instead, the kernel is handling both the read from the file and the write to the socket, so everything is happening inside the kernel.